File: src/lib/event_publisher/src/event_publisher.cpp
#include "event_publisher/event_publisher.hpp"

#include <ros/ros.h>

#include <boost/uuid/uuid.hpp>
#include <boost/uuid/uuid_generators.hpp>
#include <boost/uuid/uuid_io.hpp>
#include <chrono>
#include <mutex>

namespace open_mower::events {

namespace {

std::mutex g_mutex;
ros::Publisher g_publisher;
std::string g_source;
bool g_initialized = false;
bool g_warned_uninitialized = false;

uint64_t now_epoch_ms() {
  using namespace std::chrono;
  return static_cast<uint64_t>(duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count());
}

std::string make_uuid() {
  static thread_local boost::uuids::random_generator gen;
  return boost::uuids::to_string(gen());
}

}  // namespace

void EventPublisher::init(ros::NodeHandle& nh, const std::string& source) {
  std::lock_guard<std::mutex> lock(g_mutex);
  // Topic is global on purpose — every producer publishes onto the same bus
  // so xbot_monitoring (the single subscriber) sees them all.
  g_publisher = nh.advertise<xbot_msgs::Event>("/events", 20, /*latch=*/false);
  g_source = source;
  g_initialized = true;
}

void EventPublisher::emit(uint8_t severity, const std::string& type, const std::string& summary,
                          const nlohmann::json& details) {
  std::lock_guard<std::mutex> lock(g_mutex);
  if (!g_initialized) {
    if (!g_warned_uninitialized) {
      ROS_WARN("EventPublisher::emit called before init(); event '%s' dropped", type.c_str());
      g_warned_uninitialized = true;
    }
    return;
  }

  xbot_msgs::Event msg;
  msg.header.stamp = ros::Time::now();
  msg.id = make_uuid();
  msg.ts_ms = now_epoch_ms();
  msg.severity = severity;
  msg.type = type;
  msg.source = g_source;
  msg.summary = summary;
  msg.details_json = details.empty() ? std::string() : details.dump();

  g_publisher.publish(msg);
}

void EventPublisher::info(const std::string& type, const std::string& summary, const nlohmann::json& details) {
  emit(xbot_msgs::Event::SEVERITY_INFO, type, summary, details);
}

void EventPublisher::warning(const std::string& type, const std::string& summary, const nlohmann::json& details) {
  emit(xbot_msgs::Event::SEVERITY_WARNING, type, summary, details);
}

void EventPublisher::error(const std::string& type, const std::string& summary, const nlohmann::json& details) {
  emit(xbot_msgs::Event::SEVERITY_ERROR, type, summary, details);
}

void EventPublisher::critical(const std::string& type, const std::string& summary, const nlohmann::json& details) {
  emit(xbot_msgs::Event::SEVERITY_CRITICAL, type, summary, details);
}

}  // namespace open_mower::events

File: src/mower_scheduler/nodes/event_publisher.py
"""Python twin of open_mower::events::EventPublisher.

Producers (currently mower_scheduler) call ``init(source)`` once at node
startup, then ``emit(severity, type, summary, details=None)`` at each
lifecycle transition. Events are published on the global ``/events`` ROS
topic; ``xbot_monitoring`` is the single subscriber that persists them and
mirrors them onto MQTT.
"""

from __future__ import annotations

import json
import threading
import time
import uuid
from typing import Any, Optional

import rospy
from xbot_msgs.msg import Event

SEVERITY_INFO = Event.SEVERITY_INFO
SEVERITY_WARNING = Event.SEVERITY_WARNING
SEVERITY_ERROR = Event.SEVERITY_ERROR
SEVERITY_CRITICAL = Event.SEVERITY_CRITICAL

_lock = threading.Lock()
_publisher: Optional[rospy.Publisher] = None
_source: str = ""
_warned_uninitialized = False


def init(source: str) -> None:
    """Initialise the singleton publisher. Call once after ``rospy.init_node``."""
    global _publisher, _source
    with _lock:
        _publisher = rospy.Publisher("/events", Event, queue_size=20, latch=False)
        _source = source


def emit(severity: int, event_type: str, summary: str, details: Optional[dict[str, Any]] = None) -> None:
    """Publish a single event. Safe to call before init() — silently dropped."""
    global _warned_uninitialized
    with _lock:
        if _publisher is None:
            if not _warned_uninitialized:
                rospy.logwarn("event_publisher.emit called before init(); event '%s' dropped", event_type)
                _warned_uninitialized = True
            return
        publisher = _publisher
        source = _source

    msg = Event()
    msg.header.stamp = rospy.Time.now()
    msg.id = str(uuid.uuid4())
    msg.ts_ms = int(time.time() * 1000)
    msg.severity = int(severity)
    msg.type = event_type
    msg.source = source
    msg.summary = summary
    msg.details_json = json.dumps(details, separators=(",", ":")) if details else ""
    publisher.publish(msg)


def info(event_type: str, summary: str, details: Optional[dict[str, Any]] = None) -> None:
    emit(SEVERITY_INFO, event_type, summary, details)


def warning(event_type: str, summary: str, details: Optional[dict[str, Any]] = None) -> None:
    emit(SEVERITY_WARNING, event_type, summary, details)


def error(event_type: str, summary: str, details: Optional[dict[str, Any]] = None) -> None:
    emit(SEVERITY_ERROR, event_type, summary, details)


def critical(event_type: str, summary: str, details: Optional[dict[str, Any]] = None) -> None:
    emit(SEVERITY_CRITICAL, event_type, summary, details)
